Fix use-after-move in WebExtensionTab::captureVisibleWebpage

ddkilzer opened the pull request:
#### ea1f43afcb3f71686bf2f0db189c1385479f526c
<pre>
Fix use-after-move in WebExtensionTab::captureVisibleWebpage
<a href="https://bugs.webkit.org/show_bug.cgi?id=267823">https://bugs.webkit.org/show_bug.cgi?id=267823</a>
&lt;<a href="https://rdar.apple.com/121324738">rdar://121324738</a>&gt;

Reviewed by Timothy Hatcher.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::captureVisibleWebpage):
- Rearrange the logic in the method to move the early return that calls
  completionHandler() to the top of the method.

Canonical link: <a href="https://commits.webkit.org/273284@main">https://commits.webkit.org/273284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5fd3aa35403b88b4fe4f0bfdafb9394972aa3ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34984 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13868 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37727 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31594 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/16261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10946 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11750 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31193 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10296 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10356 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38979 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31813 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/31611 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36380 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10460 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34365 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12262 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10993 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4496 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11326 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -627,6 +627,14 @@
+    if (!canUseCaptureVisibleWebpage && !mainWebView) {
+        completionHandler(nil, toErrorString(@"tabs.captureVisibleTab()", nil, @"capture is unavailable for this tab"));
+        return;
+    }
```
I think this would read clearer as `delegateIsUnavailable`, to avoid `!` everywhere. I also don't think `__weak` is needed here, is it?

```
bool delegateIsUnavailable = !isValid() || !m_respondsToCaptureVisibleWebpage;
WKWebView *mainWebView = delegateIsUnavailable ? this->mainWebView() : nil;

if (delegateIsUnavailable && !mainWebView) {
    completionHandler(nil, toErrorString(@"tabs.captureVisibleTab()", nil, @"capture is unavailable for this tab"));
    return;
}
```

ddkilzer replied:
Done!  And `__weak` is not needed—I forgot this file is compiled with ARC—so I'll remove it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -638,13 +646,7 @@
-            return;
-        }
-
+    if (!canUseCaptureVisibleWebpage) {
```
```
if (delegateIsUnavailable) {
```

Commit:
Fix use-after-move in WebExtensionTab::captureVisibleWebpage
https://bugs.webkit.org/show_bug.cgi?id=267823
<rdar://121324738>

Reviewed by Timothy Hatcher.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::captureVisibleWebpage):
- Rearrange the logic in the method to move the early return that calls
  completionHandler() to the top of the method.

Canonical link: https://commits.webkit.org/273284@main
